executor/common_linux.h

```diff
@@ -766,6 +766,138 @@ static void initialize_devlink_pci(void)
 #endif
 #endif
 
+#if SYZ_EXECUTOR || SYZ_NIC_VF
+#include <ifaddrs.h>
+#include <linux/ethtool.h>
+#include <linux/sockios.h>
+
+struct vf_intf {
+	char pass_thru_intf[IFNAMSIZ];
+	int ppid; // used by Child
+};
+
+static struct vf_intf the_vf_struct, *vfp = &the_vf_struct;
+
+static void find_vf_interface(void)
+{
+	char cmdline[256];
+	struct ifaddrs* addresses = NULL;
+	int pid = getpid();
+	int ret = 0;
+
+	memset(vfp, 0, sizeof(struct vf_intf));
+
+	if (system("lspci | grep ^00:11.0") != 0)
+		return;
+
+	debug("Checking for VF pass-thru interface.\n");
+	if (getifaddrs(&addresses) == -1) {
+		perror("getifaddrs");
+		return;
+	}
+
+	int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+
+	if (fd < 0) {
+		perror("socket");
+		return;
+	}
+	struct ifreq ifr;
+	struct ethtool_drvinfo drvinfo;
+	struct ifaddrs* address = addresses;
+
+	while (address) {
+		debug("ifa_name: %s\n", address->ifa_name);
+		memset(&ifr, 0, sizeof(struct ifreq));
+		strcpy(ifr.ifr_name, address->ifa_name);
+		memset(&drvinfo, 0, sizeof(struct ethtool_drvinfo));
+		drvinfo.cmd = ETHTOOL_GDRVINFO;
+		ifr.ifr_data = (caddr_t)&drvinfo;
+		ret = ioctl(fd, SIOCETHTOOL, &ifr);
+
+		if (ret < 0) {
+			perror("ioctl");
+		} else if (strlen(drvinfo.bus_info)) {
+			debug("bus_info: %s, strlen(drvinfo.bus_info)=%zu\n",
+			      drvinfo.bus_info, strlen(drvinfo.bus_info));
+			if (strncmp(drvinfo.bus_info, "0000:00:11.0",
+				    strlen("0000:00:11.0")) == 0) {
+				if (strlen(address->ifa_name) <= IFNAMSIZ) {
+					strncpy(vfp->pass_thru_intf, address->ifa_name, IFNAMSIZ);
+					vfp->ppid = getpid();
+					break;
+				}
+			}
+		}
+		address = address->ifa_next;
+	}
+	freeifaddrs(addresses);
+	if (!vfp->ppid)
+		goto error;
+
+	sprintf(cmdline, "ip netns del staging_ns.%d", pid);
+	if (system(cmdline))
+		perror("system: ip netns del staging_ns");
+
+	sprintf(cmdline, "ip netns add staging_ns.%d", pid);
+	ret = system(cmdline);
+	if (ret)
+		goto error;
+	sprintf(cmdline, "ip link set dev %s netns staging_ns.%d", vfp->pass_thru_intf, pid);
```

Comment:
Why do we need to move it do staging_ns.%d, if we move it to the child process namespace in initialize_vf_passthru anyway? Can't we move it directly to the child namespace?

Reply:
As you suggested, got rid of staging namespace and nic vf interface is now moved to the syzkaller network namespace. Right after startup can dump similar per VM stats:

Reply:
sshport 59378 nsenter -t 1591 -n -- ip a show nicvf0
63: nicvf0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether aa:aa:aa:aa:aa:25 brd ff:ff:ff:ff:ff:ff
    inet 172.20.20.37/24 scope global nicvf0
       valid_lft forever preferred_lft forever
    inet6 fe80::a8aa:aaff:feaa:aa25/64 scope link dadfailed tentative 
       valid_lft forever preferred_lft forever
    inet6 fe80::25/120 scope link dadfailed tentative 
       valid_lft forever preferred_lft forever
sshport 59378 nsenter -t 1591 -n -- ip -s -s link show dev nicvf0
63: nicvf0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether aa:aa:aa:aa:aa:25 brd ff:ff:ff:ff:ff:ff
    RX: bytes  packets  errors  dropped overrun mcast   
    5400       61       0       0       0       58      
    RX errors: length   crc     frame   fifo    missed
               0        0       0       0       0       
    TX: bytes  packets  errors  dropped carrier collsns 
    612        6        0       0       0       0       
    TX errors: aborted  fifo   window heartbeat transns
               0        0       0       0       4       
sshport 59378 nsenter -t 1591 -n -- ethtool -i nicvf0
driver: bnxt_en
version: 5.19.0-rc7-syzk
firmware-version: 20.8.177.0/pkg 20.8.29.0
expansion-rom-version: 
bus-info: 0000:00:11.0
supports-statistics: yes
supports-test: no
supports-eeprom-access: no
supports-register-dump: no
supports-priv-flags: no